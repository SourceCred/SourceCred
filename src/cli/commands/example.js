// @flow
import {Command, flags} from "@oclif/command";

export default class ExampleCommand extends Command {
  static description = "An example command description";
  static flags = {
    name: flags.string({char: "n", description: "name to print"}),
  };
  async run() {
    const {flags} = this.parse(ExampleCommand);
    const name = flags.name || "world";
    this.log(`hello ${name} EXAMPEL`);
  }
}
