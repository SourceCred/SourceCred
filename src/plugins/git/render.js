// @flow

import type {Address} from "core/address";
import type {Graph} from "core/graph";
import type {NodeType, SubmoduleCommitPayload} from "./types";

/**
 * Describe a node provided by the Git plugin.
 */
export function nodeDescription(graph: Graph<any, any>, address: Address) {
  const type: NodeType = (address.type: any);
  switch (type) {
    case "COMMIT":
      return `commit ${address.id}`;
    case "TREE":
      return `tree ${address.id}`;
    case "BLOB":
      return `blob ${address.id}`;
    case "SUBMODULE_COMMIT": {
      const payload: SubmoduleCommitPayload = graph.node(address).payload;
      return `submodule commit ${payload.hash} in ${payload.url}`;
    }
    case "TREE_ENTRY":
      return `entry ${address.id}`;
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`unknown type: ${type}`);
  }
}
