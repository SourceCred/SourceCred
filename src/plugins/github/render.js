// @flow

import React, {type Node as ReactNode} from "react";
import * as R from "./relationalView";

function EntityUrl(props) {
  return (
    <a href={props.entity.url()} target="_blank" rel="nofollow noopener">
      {props.children}
    </a>
  );
}

function repo(x: R.Repo) {
  return (
    <EntityUrl entity={x}>
      {x.owner()}/{x.name()}
    </EntityUrl>
  );
}

function hyperlinkedNumber(x: R.Issue | R.Pull) {
  return <EntityUrl entity={x}>#{x.number()}</EntityUrl>;
}

function issue(x: R.Issue) {
  return (
    <span>
      {hyperlinkedNumber(x)}: {x.title()}
    </span>
  );
}

function pull(x: R.Pull) {
  const additions = <span style={{color: "green"}}>+{x.additions()}</span>;
  const deletions = <span style={{color: "red"}}>âˆ’{x.deletions()}</span>;
  const diff = (
    <span>
      ({additions}/{deletions})
    </span>
  );
  return (
    <span>
      {hyperlinkedNumber(x)} {diff}: {x.title()}
    </span>
  );
}

function review(x: R.Review) {
  const leader = <EntityUrl entity={x}>review</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function comment(x: R.Comment) {
  const leader = <EntityUrl entity={x}>comment</EntityUrl>;
  return (
    <span>
      {leader} on {description(x.parent())}
    </span>
  );
}

function userlike(x: R.Userlike) {
  return <EntityUrl entity={x}>@{x.login()}</EntityUrl>;
}

// The commit type is included for completeness's sake and to
// satisfy the typechecker, but won't ever be seen in the frontend
// because the commit has a Git plugin prefix and will therefore by
// handled by the git plugin adapter
function commit(x: R.Commit) {
  // TODO(@wchargin): Ensure this hash is unambiguous
  const shortHash = x.address().hash.slice(0, 7);
  return (
    <span>
      Commit <EntityUrl entity={x}>{shortHash}</EntityUrl>
    </span>
  );
}

export function description(e: R.Entity): ReactNode {
  const handlers = {
    repo,
    issue,
    pull,
    review,
    comment,
    commit,
    userlike,
  };
  return R.match(handlers, e);
}
