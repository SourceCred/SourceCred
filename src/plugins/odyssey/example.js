// @flow

import {OdysseyInstance, type Node} from "./instance";

/**
 * An example Odyssey instance, based on work at the Odyssey Hackathon.
 */
export function hackathonExample(): OdysseyInstance {
  // Define the types of nodes allowed in our instance
  const instance = new OdysseyInstance();

  // define our values for the hackathon
  const logistics = instance.addNode("VALUE", "logistics");
  const design = instance.addNode("VALUE", "design");
  const narrative = instance.addNode("VALUE", "narrative");
  const prototype = instance.addNode("VALUE", "prototype");
  const outreach = instance.addNode("VALUE", "outreach");

  // the cast of characters
  const dl = instance.addNode("PERSON", "dandelion");
  const mz = instance.addNode("PERSON", "z zargham");
  const irene = instance.addNode("PERSON", "irene");
  const max = instance.addNode("PERSON", "max");
  const dennis = instance.addNode("PERSON", "dennis");
  const jonathan = instance.addNode("PERSON", "jonathan");
  const lb = instance.addNode("PERSON", "lb");
  const brian = instance.addNode("PERSON", "brian");
  const sarah = instance.addNode("PERSON", "sarah");
  const jmnemo = instance.addNode("PERSON", "@jmnemo");
  const talbott = instance.addNode("PERSON", "jonathan talbott");
  const agata = instance.addNode("PERSON", "agata");

  // the artifacts
  const graphviz = instance.addNode("ARTIFACT", "graph visualizer");
  const backend = instance.addNode("ARTIFACT", "backend");
  const frontend = instance.addNode("ARTIFACT", "frontend");
  const seededPagerank = instance.addNode("ARTIFACT", "seeded pagerank");
  const canvas = instance.addNode(
    "ARTIFACT",
    "the awesome illustrated poster board"
  );
  const logo = instance.addNode("ARTIFACT", "the broken-lightbulb logo");

  instance.addEdge("DEPENDS_ON", prototype, graphviz);
  instance.addEdge("DEPENDS_ON", prototype, backend);
  instance.addEdge("DEPENDS_ON", prototype, frontend);
  instance.addEdge("DEPENDS_ON", frontend, seededPagerank);
  instance.addEdge("DEPENDS_ON", design, graphviz);
  instance.addEdge("DEPENDS_ON", design, frontend);
  instance.addEdge("DEPENDS_ON", design, logo);
  instance.addEdge("DEPENDS_ON", narrative, logo);
  instance.addEdge("DEPENDS_ON", narrative, canvas);

  function addContribution(
    description: string,
    authors: Node[],
    impacted: Node[]
  ) {
    const contrib = instance.addNode("CONTRIBUTION", description);
    for (const author of authors) {
      instance.addEdge("DEPENDS_ON", contrib, author);
    }
    for (const impact of impacted) {
      instance.addEdge("DEPENDS_ON", impact, contrib);
    }
  }

  addContribution(
    "colors for the graph visualizer",
    [dennis, irene, lb, max, dl],
    [design, graphviz]
  );
  addContribution("design for graph visualizer", [dennis], [graphviz, design]);
  addContribution(
    "design for the frontend",
    [dennis, irene],
    [design, frontend]
  );
  addContribution(
    "pre-hack planning and project management",
    [brian],
    [prototype, logistics]
  );
  addContribution("implementing the graph visualizer", [dl], [graphviz]);
  addContribution("implementing seeded PageRank", [dl, mz], [seededPagerank]);
  addContribution("implementing the frontend", [dl, jmnemo], [frontend]);
  addContribution("implementing the backend", [dl], [backend]);
  addContribution("logo--preliminary work", [lb, agata], [logo]);
  addContribution("logo--lightbulb moment", [lb, max], [logo]);
  addContribution("drawing the canvas", [lb], [canvas]);
  addContribution(
    "narrative shaping for canvas",
    [lb, dl, mz],
    [canvas, narrative]
  );
  addContribution("oneline narrative statement", [dl, talbott], [narrative]);
  addContribution("oneline narrative review", [max, dl], [narrative]);
  addContribution("booking hotel stay", [sarah], [logistics]);
  addContribution("booking plane tickets", [sarah], [logistics]);
  addContribution("helping Dennis get into the space", [sarah], [logistics]);
  addContribution(
    "example prototype dataset",
    [mz, brian],
    [prototype, outreach]
  );
  addContribution("final presentation", [dl, prototype], [narrative]);
  addContribution("connection to the common stack", [mz], [outreach]);
  addContribution("discussing the canvas with folks", [lb], [outreach]);
  addContribution("general logistical defense", [jonathan], [logistics]);
  addContribution("recruiting Max & company", [mz], [outreach, logistics]);
  addContribution("forming the team", [mz], [logistics]);

  return instance;
}
