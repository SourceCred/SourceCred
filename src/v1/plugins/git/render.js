// @flow

import {NodeReference} from "../../core/porcelain";
import {SubmoduleCommitReference, GitReference} from "./porcelain";

/**
 * Describe a node provided by the Git plugin.
 */
export function nodeDescription(nodeReference: NodeReference<any>) {
  const gReference = new GitReference(nodeReference);
  const type = gReference.type();
  const address = gReference.address();
  switch (type) {
    case "COMMIT":
      return `commit ${address.id}`;
    case "TREE":
      return `tree ${address.id}`;
    case "BLOB":
      return `blob ${address.id}`;
    case "SUBMODULE_COMMIT": {
      const scRef = new SubmoduleCommitReference(gReference);
      const scPorcelain = scRef.get();
      if (scPorcelain != null) {
        return `submodule commit ${scPorcelain.hash()} in ${scPorcelain.url()}`;
      } else {
        return `submodule commit [unknown]`;
      }
    }
    case "TREE_ENTRY":
      return `entry ${address.id}`;
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(`unknown type: ${type}`);
  }
}
