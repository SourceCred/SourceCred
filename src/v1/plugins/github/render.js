// @flow

/*
 * Methods for rendering and displaying GitHub nodes.
 */
import stringify from "json-stable-stringify";
import type {NodeReference} from "../../core/porcelain";
import type {NodePayload} from "./types";
import {
  GithubReference,
  AuthorReference,
  AuthorPorcelain,
  CommentReference,
  IssueReference,
  IssuePorcelain,
  PullRequestReference,
  PullRequestPorcelain,
  PullRequestReviewReference,
  PullRequestReviewCommentReference,
  RepositoryPorcelain,
} from "./porcelain";

/* Give a short description for the GitHub node at given address.
 * Useful for e.g. displaying a title.
 */
export function nodeDescription(ref: NodeReference<NodePayload>) {
  const porcelain = ref.get();
  if (porcelain == null) {
    return `[unknown ${ref.type()}]`;
  }
  const type = new GithubReference(ref).type();
  switch (type) {
    case "REPOSITORY": {
      const repo = new RepositoryPorcelain(porcelain);
      return `${repo.owner()}/${repo.name()}`;
    }
    case "ISSUE": {
      const issue = new IssuePorcelain(porcelain);
      return `#${issue.number()}: ${issue.title()}`;
    }
    case "PULL_REQUEST": {
      const pr = new PullRequestPorcelain(porcelain);
      const diff = `+${pr.additions()}/\u2212${pr.deletions()}`;
      return `#${pr.number()} (${diff}): ${pr.title()}`;
    }
    case "COMMENT": {
      const comment = new CommentReference(ref);
      const issue = comment.parent();
      return `comment by @${authors(comment)} on #${num(issue)}`;
    }
    case "PULL_REQUEST_REVIEW": {
      const review = new PullRequestReviewReference(ref);
      const pr = review.parent();
      return `review by @${authors(review)} on #${num(pr)}`;
    }
    case "PULL_REQUEST_REVIEW_COMMENT": {
      const comment = new PullRequestReviewCommentReference(ref);
      const pr = comment.parent().parent();
      return `review comment by @${authors(comment)} on #${num(pr)}`;
    }
    case "AUTHOR": {
      return `@${new AuthorPorcelain(porcelain).login()}`;
    }
    default: {
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw new Error(
        `Tried to write description for invalid type ${stringify(
          ref.address()
        )}`
      );
    }
  }
}

function num(x: IssueReference | PullRequestReference) {
  const np = x.get();
  return np == null ? "[unknown]" : np.number();
}

function authors(authorable: {+authors: () => AuthorReference[]}) {
  // TODO: modify to accomodate multi-authorship
  const authorRefs = authorable.authors();
  const firstAuthor = authorRefs[0].get();
  return firstAuthor != null ? firstAuthor.login() : "[unknown]";
}
