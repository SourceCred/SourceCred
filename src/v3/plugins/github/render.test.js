// @flow

import {exampleEntities} from "./example/example";
import {description} from "./render";

describe("plugins/github/render", () => {
  it("descriptions are as expected", () => {
    const examples = exampleEntities();
    const withDescriptions = {};
    for (const name of Object.keys(exampleEntities())) {
      const entity = examples[name];
      withDescriptions[name] = description(entity);
    }
    expect(withDescriptions).toMatchSnapshot();
  });
});
