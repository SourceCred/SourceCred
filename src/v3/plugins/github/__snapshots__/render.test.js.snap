// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/render descriptions are as expected 1`] = `
Object {
  "comment": "comment by @wchargin on review by @wchargin of #5: This pull request will be more contentious. I can feel it...",
  "issue": "#2: A referencing issue.",
  "pull": "#5: This pull request will be more contentious. I can feel it...",
  "repo": "sourcecred/example-github",
  "review": "review by @wchargin of #5: This pull request will be more contentious. I can feel it...",
  "userlike": "@wchargin",
}
`;
